File: plugins/cognos-to-sigma/skills/cognos-to-sigma/scripts/lib/warehouse-column-refs.mjs
const norm = (value) => String(value ?? '').toLowerCase().replace(/[^a-z0-9]/g, '');

function catalogMatch(entries, candidates) {
  const names = entries.map((entry) => String(entry.name || '')).filter(Boolean);
  for (const candidate of candidates.filter(Boolean).map(String)) {
    if (names.includes(candidate)) return candidate;
    const folded = names.filter((name) => name.toLowerCase() === candidate.toLowerCase());
    if (folded.length === 1) return folded[0];
    const normalized = names.filter((name) => norm(name) === norm(candidate));
    if (normalized.length === 1) return normalized[0];
  }
  return null;
}

async function jsonApi(api, method, path, body) {
  const response = await api(method, path, body);
  if (!response.ok || !response.json) throw new Error(`${method} ${path} failed (HTTP ${response.status}): ${response.text.slice(0, 300)}`);
  return response.json;
}

async function listColumns(api, tableId) {
  const entries = [];
  let token = null;
  do {
    const query = new URLSearchParams({ pageSize: '500' });
    if (token) query.set('pageToken', token);
    const page = await jsonApi(api, 'GET', `/v2/connections/tables/${tableId}/columns?${query}`);
    entries.push(...(page.entries || []));
    token = page.nextPageToken || null;
  } while (token);
  return entries;
}

export async function groundWarehouseRefs(spec, api) {
  const modes = new Map(), cache = new Map(), aliases = new Map(), prefixes = new Set(), idMap = new Map();
  const unresolved = [];
  let rewritten = 0, rekeyed = 0, reprefixed = 0;
  const elements = (spec.pages || []).flatMap((page) => page.elements || []);
  for (const element of elements) {
    const source = element.source || {};
    if (source.kind !== 'warehouse-table') continue;
    const connection = String(source.connectionId || ''), path = source.path;
    if (!connection || !Array.isArray(path) || !path.length) continue;
    if (!modes.has(connection)) {
      const details = await jsonApi(api, 'GET', `/v2/connections/${connection}`);
      modes.set(connection, details.friendlyName);
    }
    const friendly = modes.get(connection);
    if (friendly !== true && friendly !== false) throw new Error(`connection ${connection} did not report friendlyName`);
    if (friendly) continue;

    const oldName = String(element.name || ''), physicalName = String(path.at(-1));
    if (oldName && oldName !== physicalName) aliases.set(oldName, physicalName);
    element.name = physicalName;
    const cacheKey = `${connection}\0${JSON.stringify(path)}`;
    if (!cache.has(cacheKey)) {
      const table = await jsonApi(api, 'POST', `/v2/connection/${connection}/lookup`, { path });
      if (table.kind !== 'table' || !table.inodeId) throw new Error(`${path.join('.')} did not resolve to a table`);
      cache.set(cacheKey, await listColumns(api, table.inodeId));
    }
    const entries = cache.get(cacheKey);
    const refs = new Map(entries.filter((entry) => entry.name).map((entry) => [norm(entry.name), String(entry.name)]));
    prefixes.add(oldName); prefixes.add(physicalName);

    for (const column of element.columns || []) {
      const match = String(column.formula || '').match(/^\[([^\]/]+)\/([^\]/]+)\]$/);
      if (!match) continue;
      const [, prefix, leaf] = match;
      const cid = String(column.id || ''), idLeaf = cid.includes('/') ? cid.split('/', 2)[1] : null;
      const physical = catalogMatch(entries, [leaf, column.name, idLeaf]);
      if (!physical) {
        if (cid.startsWith('inode-')) unresolved.push(`${path.join('.')}: ${prefix}/${leaf}`);
        continue;
      }
      if (!column.name) column.name = leaf;
      if (cid.startsWith('inode-') && idLeaf !== physical) {
        const newId = `${cid.split('/', 1)[0]}/${physical}`;
        idMap.set(cid, newId); column.id = newId; rekeyed++;
      }
      const grounded = `[${physicalName}/${physical}]`;
      if (column.formula !== grounded) { column.formula = grounded; rewritten++; }
    }

    const sameElement = (node) => {
      if (Array.isArray(node)) return node.forEach(sameElement);
      if (!node || typeof node !== 'object') return;
      let display = null;
      if (typeof node.formula === 'string') {
        node.formula = node.formula.replace(/\[([^\]/]+)(\/[^\]]+)\]/g, (whole, prefix, suffix) => {
          const leaf = suffix.slice(1);
          if (leaf.includes('/') || ![oldName, physicalName].includes(prefix) || !refs.has(norm(leaf))) return whole;
          display ||= leaf;
          const grounded = `[${physicalName}/${refs.get(norm(leaf))}]`;
          if (grounded !== whole) rewritten++;
          return grounded;
        });
      }
      if (display && !node.name) node.name = display;
      for (const [key, value] of Object.entries(node)) if (key !== 'formula') sameElement(value);
    };
    sameElement(element);
  }

  const replaceIds = (node) => {
    if (Array.isArray(node)) return node.map(replaceIds);
    if (node && typeof node === 'object') {
      for (const [key, value] of Object.entries(node)) node[key] = replaceIds(value);
      return node;
    }
    return typeof node === 'string' ? (idMap.get(node) || node) : node;
  };
  replaceIds(spec);

  const derived = (node) => {
    if (Array.isArray(node)) return node.forEach(derived);
    if (!node || typeof node !== 'object') return;
    let display = null;
    if (typeof node.formula === 'string') {
      node.formula = node.formula.replace(/\[([^\]/]+)(\/[^\]]+)\]/g, (_whole, prefix, suffix) => {
        const replacement = aliases.get(prefix) || prefix;
        if (replacement !== prefix) reprefixed++;
        if (!suffix.slice(1).includes('/') && (prefixes.has(prefix) || prefixes.has(replacement))) display ||= suffix.slice(1);
        return `[${replacement}${suffix}]`;
      });
    }
    if (display && !node.name) node.name = display;
    for (const [key, value] of Object.entries(node)) if (key !== 'formula') derived(value);
  };
  derived(spec);
  if (unresolved.length) throw new Error(`warehouse columns not found in catalog: ${[...new Set(unresolved)].join(', ')}`);
  return { rewritten, rekeyed, reprefixed, connectionModes: Object.fromEntries(modes) };
}

File: plugins/thoughtspot-to-sigma/skills/thoughtspot-to-sigma/scripts/warehouse_column_refs.py
"""Ground warehouse-table references for Sigma connections without friendly names."""
import re
import urllib.parse


def _norm(value):
    return re.sub(r"[^a-z0-9]", "", str(value or "").lower())


def _match(entries, candidates):
    names = [str(entry.get("name") or "") for entry in entries if entry.get("name")]
    for candidate in [str(value) for value in candidates if value]:
        if candidate in names:
            return candidate
        folded = [name for name in names if name.lower() == candidate.lower()]
        if len(folded) == 1:
            return folded[0]
        normalized = [name for name in names if _norm(name) == _norm(candidate)]
        if len(normalized) == 1:
            return normalized[0]
    return None


def _columns(api, path):
    out, token = [], None
    while True:
        query = {"pageSize": 500}
        if token:
            query["pageToken"] = token
        page = api("GET", path + "?" + urllib.parse.urlencode(query))
        out.extend(page.get("entries") or [])
        token = page.get("nextPageToken")
        if not token:
            return out


def apply(spec, api):
    modes, cache, aliases, prefixes, id_map, unresolved = {}, {}, {}, set(), {}, []
    rewritten = rekeyed = reprefixed = 0
    elements = [e for p in (spec.get("pages") or []) for e in (p.get("elements") or [])]
    for element in elements:
        source = element.get("source") or {}
        if source.get("kind") != "warehouse-table":
            continue
        connection, path = str(source.get("connectionId") or ""), source.get("path")
        if not connection or not isinstance(path, list) or not path:
            continue
        if connection not in modes:
            modes[connection] = api("GET", f"/v2/connections/{connection}").get("friendlyName")
        friendly = modes[connection]
        if friendly not in (True, False):
            raise RuntimeError(f"connection {connection} did not report friendlyName")
        if friendly:
            continue
        old, physical_name = str(element.get("name") or ""), str(path[-1])
        if old and old != physical_name:
            if old in aliases and aliases[old] != physical_name:
                raise RuntimeError(f"ambiguous warehouse element name {old!r}")
            aliases[old] = physical_name
        element["name"] = physical_name
        key = (connection, tuple(path))
        if key not in cache:
            table = api("POST", f"/v2/connection/{connection}/lookup", {"path": path})
            if table.get("kind") != "table" or not table.get("inodeId"):
                raise RuntimeError(f"{'.'.join(path)} did not resolve to a table")
            cache[key] = _columns(api, f"/v2/connections/tables/{table['inodeId']}/columns")
        entries = cache[key]
        refs = {_norm(entry.get("name")): str(entry.get("name")) for entry in entries if entry.get("name")}
        prefixes.update(filter(None, (old, physical_name)))
        for column in element.get("columns") or []:
            match = re.fullmatch(r"\[([^\]/]+)/([^\]/]+)\]", str(column.get("formula") or ""))
            if not match:
                continue
            prefix, leaf = match.groups()
            cid = str(column.get("id") or "")
            id_leaf = cid.split("/", 1)[1] if "/" in cid else None
            physical = _match(entries, (leaf, column.get("name"), id_leaf))
            if not physical:
                if cid.startswith("inode-"):
                    unresolved.append(f"{'.'.join(path)}: {prefix}/{leaf}")
                continue
            column.setdefault("name", leaf)
            if cid.startswith("inode-") and id_leaf != physical:
                new_id = cid.split("/", 1)[0] + "/" + physical
                id_map[cid] = new_id
                column["id"] = new_id
                rekeyed += 1
            grounded = f"[{physical_name}/{physical}]"
            if column.get("formula") != grounded:
                column["formula"] = grounded
                rewritten += 1

        def same_element(node):
            nonlocal rewritten
            if isinstance(node, dict):
                display = None
                if isinstance(node.get("formula"), str):
                    def replace(match):
                        nonlocal display, rewritten
                        prefix, leaf = match.group(1), match.group(2)[1:]
                        if "/" in leaf or prefix not in (old, physical_name) or _norm(leaf) not in refs:
                            return match.group(0)
                        display = display or leaf
                        grounded = f"[{physical_name}/{refs[_norm(leaf)]}]"
                        rewritten += grounded != match.group(0)
                        return grounded
                    node["formula"] = re.sub(r"\[([^\]/]+)(/[^\]]+)\]", replace, node["formula"])
                if display and not node.get("name"):
                    node["name"] = display
                for key2, value in node.items():
                    if key2 != "formula":
                        same_element(value)
            elif isinstance(node, list):
                for value in node:
                    same_element(value)
        same_element(element)

    def replace_ids(node):
        if isinstance(node, dict):
            for key, value in list(node.items()):
                node[key] = replace_ids(value)
        elif isinstance(node, list):
            for index, value in enumerate(node):
                node[index] = replace_ids(value)
        elif isinstance(node, str):
            return id_map.get(node, node)
        return node
    replace_ids(spec)

    def derived(node):
        nonlocal reprefixed
        if isinstance(node, dict):
            display = None
            if isinstance(node.get("formula"), str):
                def replace(match):
                    nonlocal display, reprefixed
                    prefix, suffix = match.group(1), match.group(2)
                    replacement = aliases.get(prefix, prefix)
                    reprefixed += replacement != prefix
                    if "/" not in suffix[1:] and (prefix in prefixes or replacement in prefixes):
                        display = display or suffix[1:]
                    return f"[{replacement}{suffix}]"
                node["formula"] = re.sub(r"\[([^\]/]+)(/[^\]]+)\]", replace, node["formula"])
            if display and not node.get("name"):
                node["name"] = display
            for key, value in node.items():
                if key != "formula":
                    derived(value)
        elif isinstance(node, list):
            for value in node:
                derived(value)
    derived(spec)
    if unresolved:
        raise RuntimeError("warehouse columns not found in catalog: " + ", ".join(sorted(set(unresolved))))
    return {"rewritten": rewritten, "rekeyed": rekeyed, "reprefixed": reprefixed,
            "connectionModes": modes}
